### PR TITLE
dm: display instruction separately if precheck error is terror.Error

### PR DESCRIPTION
**Pull request by D3Hunter**
### What problem does this PR solve?

Issue Number: ref #4287

#### Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

### Release note 

```release-note
`None`.
```

### CODE REVIEW

File: dm/pkg/checker/utils.go [hunk 1]
```diff
@@ -135,7 +136,12 @@ func markCheckError(result *Result, err error) {
 		if result.State != StateFailure {
 			result.State = state
 		}
-		result.Errors = append(result.Errors, &Error{Severity: state, ShortErr: err.Error()})
+		if err2, ok := err.(*terror.Error); ok {
+			result.Errors = append(result.Errors, &Error{Severity: state, ShortErr: err2.ErrorWithoutWorkaround()})
+			result.Instruction = err2.Workaround()
```
Comment (lance6716) [inline dm/pkg/checker/utils.go:141]: How does it look like?

Comment (D3Hunter) [reply dm/pkg/checker/utils.go:141]: image

Comment (buchuitoudegou) [reply dm/pkg/checker/utils.go:141]: Can we trim the `[....],` from the `short_error`? Looks confusing.

File: dm/pkg/terror/terror.go [hunk 2]
```diff
@@ -204,15 +204,21 @@ func (e *Error) Workaround() string {
 	return e.workaround
 }
 
-// Error implements error interface.
-func (e *Error) Error() string {
+// ErrorWithoutWorkaround returns err msg without workaround, in some place like cloud we want display it separately.
```
Comment (buchuitoudegou) [inline dm/pkg/terror/terror.go:207]: I remember the `instruction` fields have been filled for all the check items (i.e., `Results[i].Instruction`). Where can we use this function?

Comment (D3Hunter) [reply dm/pkg/terror/terror.go:207]: if precheck fail with err, there is no instruction. it's used in `markCheckError`

File: dm/pkg/checker/utils.go (hunk 1 above)
Comment (D3Hunter) [reply dm/pkg/checker/utils.go:141]: https://github.com/pingcap/tiflow/pull/8201/commits/4f082b8641c325ddc4c7f15ffa117f0fab423e57

image

File: dm/pkg/terror/terror.go (hunk 2 above)
Comment (buchuitoudegou) [reply dm/pkg/terror/terror.go:207]: https://github.com/pingcap/tiflow/blob/a2d2a5213c3963485e5cb9f3bd7b6967302b9bb6/dm/pkg/checker/privilege.go#L110-L114

The `instruction` is set in the `Result` and `dm-on-cloud` uses this field as the `solution`. Do you mean it's only used in op?

**Review by buchuitoudegou — APPROVED** [review]

File: dm/pkg/checker/utils.go (hunk 1 above)
Comment (lance6716) [reply dm/pkg/checker/utils.go:141]: The `Errors` may contain multiple errors, so instruction should also be prepraed for multiple lines?

**Review by lance6716 — COMMENTED** [review]
rest lgtm

File: dm/pkg/checker/utils.go (hunk 1 above)
Comment (D3Hunter) [reply dm/pkg/checker/utils.go:141]: precheck item returns on first error meet. if error passed into it is a multi-err instance(tables check), use the `else` branch.

**Review by lance6716 — APPROVED** [review]

**Comment by D3Hunter** [comment]
6.6 branch created, will merge it

**Comment by lance6716** [comment]
some unit tests are failed @D3Hunter 

```
=== FAIL: dm/pkg/terror TestTError (0.00s)

    terror_test.go:59: 

        	Error Trace:	/home/jenkins/agent/workspace/pingcap/tiflow/ghpr_verify/tiflow/dm/pkg/terror/terror_test.go:59

        	Error:      	Not equal: 

        	            	expected: "[code=10002:class=database:scope=upstream:level=medium], Message: bad connection, Workaround: please check your network connection"

        	            	actual  : "[code=10002:class=database:scope=upstream:level=medium]Message: bad connection, Workaround: please check your network connection"

        	            	

        	            	Diff:

        	            	--- Expected

        	            	+++ Actual

        	            	@@ -1 +1 @@

        	            	-[code=10002:class=database:scope=upstream:level=medium], Message: bad connection, Workaround: please check your network connection

        	            	+[code=10002:class=database:scope=upstream:level=medium]Message: bad connection, Workaround: please check your network connection

        	Test:       	TestTError
```

**Pull request merged**